src/System.Drawing.Common/src/System/Drawing/Icon.Unix.cs

```diff
@@ -709,77 +709,163 @@ private void InitFromStreamWithSize(Stream stream, int width, int height)
+                            valid = true;
+                            BitmapInfoHeader bih = new BitmapInfoHeader
+                            {
+                                biSize = headerSize,
+                                biWidth = bihReader.ReadInt32(),
+                                biHeight = bihReader.ReadInt32(),
+                                biPlanes = bihReader.ReadUInt16(),
+                                biBitCount = bihReader.ReadUInt16(),
+                                biCompression = bihReader.ReadUInt32(),
+                                biSizeImage = bihReader.ReadUInt32(),
+                                biXPelsPerMeter = bihReader.ReadInt32(),
+                                biYPelsPerMeter = bihReader.ReadInt32(),
+                                biClrUsed = bihReader.ReadUInt32(),
+                                biClrImportant = bihReader.ReadUInt32()
+                            };
+
+                            IconImage iidata = new IconImage
+                            {
+                                iconHeader = bih
+                            };
 
-            iconDir.idType = reader.ReadUInt16();
-            if (iconDir.idType != 1) //must be 1
-                throw new System.ArgumentException("Invalid Argument", nameof(stream));
+                            // Read the number of colors used and corresponding memory occupied by
+                            // color table. Fill this memory chunk into rgbquad[]
+                            // number of colors = 2 ^ biBitCount
+                            int numColors = 0;
+                            if (bih.biBitCount < 24)
+                                numColors = (int)(1 << bih.biBitCount);
 
-            ushort dirEntryCount = reader.ReadUInt16();
-            imageData = new ImageData[dirEntryCount];
-            iconDir.idCount = dirEntryCount;
-            iconDir.idEntries = new IconDirEntry[dirEntryCount];
-            bool sizeObtained = false;
-            // now read in the IconDirEntry structures
-            for (int i = 0; i < dirEntryCount; i++)
-            {
-                IconDirEntry ide;
-                ide.width = reader.ReadByte();
-                ide.height = reader.ReadByte();
-                ide.colorCount = reader.ReadByte();
-                ide.reserved = reader.ReadByte();
-                ide.planes = reader.ReadUInt16();
-                ide.bitCount = reader.ReadUInt16();
-                ide.bytesInRes = reader.ReadUInt32();
-                ide.imageOffset = reader.ReadUInt32();
-
-                // Vista 256x256 icons points directly to a PNG bitmap
-                // 256x256 icons are decoded as 0x0 (width and height are encoded as BYTE)
-                // and we ignore them just like MS does (at least up to fx 2.0) 
-                // Added: storing data so it can be saved back
-                if ((ide.width == 0) && (ide.height == 0))
-                    ide.ignore = true;
```

Comment:
The reason why I moved it is because we where doing 2 for loops on the same data. So I saw it was possible to get the bitmapheaders on the same loop when we’re getting the directory entries instead of having another for loop and accessing the array every time to get the IconDirectoryEntry, but instead use what we just parsed. 

I can definitely move it back or split it into another commit, I tested it with various different icons and it parses and saves them correctly, but if you don’t feel confident enough and don’t think is worth the effort, I’ll leave it as it was and just add the fix without moving code around.

Reply:
Its fine. In future I would keep separate commits for refactorings. You can always squash later

Reply:
It’s easy to split it. I was just wondering if you were fine on taking this refactor or it is not worth. If you think is a nice to have I’ll split it.